settings: Add notification settings checkboxes for wildcard mentions.

timabbott opened the pull request:
This change makes it possible for users to control the notification
settings for wildcard mentions as a separate control from PMs and
direct @-mentions.

This commit is mostly a lot of plumbing similar to existing features.

-----------------------------------------------------------------------

This PR needs work in a few places:
* Strings.  @rishig thoughts on how to name this block (both before and after https://github.com/zulip/zulip/issues/12182)?  
* Wasted work.  The structure of `wildcard_mention_email_user_ids` is wrong; in most cases we're going to be sending the list of every user ID in the organization.  We should probably invert the meaning (`wildcard_mention_nopush_user_ids`) for those values, to avoid unnecessarily sending large objects for rare cases.  Shouldn't be hard to rework that and plumb it through.
* The `test_users.py` tests need to be extended to test the new code paths.
* There's another "add more unit tests" TODO as well
* Manual testing, which I can help with once we have the tests done.  Our existing test suite is pretty solid.

Some lower-value things we might want to do if easy:
* Ideally, we wouldn't have the `actions.py` code even both querying the `wildcard_mention` parameters for a message in `get_recipient_info` unless the message plausibly contains a wildcard mention, just to avoid unnecessary work.  This is made awkward by the ordering between markdown rendering and `get_recipient_info`; might not be feasible to improve as a result?
* Following this, we'll want to tackle #13096, #13080, etc.; which will certainly be nicer to work on once this is done.

@mateuszmandera this might be a good PR for you to take over to get familiarity with this code path; we have a bundle of a handful of fairly high priority notifications-type issues in this area (see https://github.com/zulip/zulip/issues?q=is%3Aopen+is%3Aissue+label%3A%22area%3A+notifications+%28messages%29%22) and it'd be nice to just tackle the bundle in a quick series over the next month or so.

timabbott commented:
![image](https://user-images.githubusercontent.com/2746074/64212813-626c7200-ce5f-11e9-8f56-94372cd73978.png)

mateuszmandera commented:
I'll start getting into this :slightly_smiling_face:

rishig commented:
Should this be a binary setting, e.g. "count `@all` mentions as mentions"? 
I think we'll also eventually want to surface this option at the stream level, and I think adding 4 more stream level settings may not make sense for this.

mateuszmandera commented:
What are the plans regarding group mentions? Because if we try to have such settings for those too, then we would really be having a lot of clutter in mention settings.
Perhaps a binary setting is more sensible :thinking:  And how many users are going to need such detailed customization - I think that most people will just want to either treat wildcard mentions like direct mentions, or if they hate getting pinged by them, to just get rid of them.

timabbott commented:
@mateuszmandera the plan is to move to a grid-like view for these settings, which should be nice and compact.  I think I mentioned that issue on this thread. 

@rishig I think that's a fair question.  I think the additional flexibility this provides is valuable.  I've definitely talked to folks who e.g. want `@all` in the `Mentions` narrow and unread counts, but don't want an email notification.  (E.g. because they check the organization weekly and only want to be called there if someone needs their personal attention, but when they do visit, they want to have things like `@all` mentions highlighted).  And I think it's more likely that long term, we'll just want a "Notification settings" checkbox for whether to include the category in the "favicon count" in our table format than that we'll want to bundle this in with the rest.

UI-wise, we can always try to simplify by making you click an "advanced" thing to see the separate settings for "Wildcard mentions", "Alert words", etc., or a "copy from mentions" button or whatever if we feel we need to make things simpler.

rishig commented:
If we're going to add this additional complexity I think it's worth mapping out what our final stream settings and notification settings page is going to look like. It's certainly possible to build simple and advanced settings pages, but I'm not sure when we'll have the resources for that given that even getting to the checkbox stage has taken a while. 

For context I wouldn't be surprised if we already have more knobs and a more complicated settings page than our competitors, and I think we're currently missing several of the more important ones.

Re: Mateusz's question: I was actually thinking that group mentions would always act as individual mentions, i.e. you wouldn't be able to configure them differently.

timabbott commented:
Well, personally I think we just want the grid view with a separate row for "wildcard mentions" from "PMs and mentions"; I think it's a significantly cleaner UI to have a column for that sort of than it is to have a checkbox somewhere else on the page that controls the behavior of that thing.  And I think a grid with like 5 rows and 5 columns will feel pretty clean and usable, and we won't end up wanting the simple/advanced pages.  Just was mentioning what directions could exist is we changed our mind.  (And what I was imagining was just the "Advanced" unhiding some hidden columns, similar to what we have on "settings" overall, if we feel the need for it).  But I think we won't really know until we've switched to the compact grid view.  Which I think @shubhamdhama is working on; I don't think it'll be super hard to do that migration.

I agree group mentions should work like individual mentions.

rishig commented:
How about in stream settings?
(I think the main configuration users will actually want is at the stream level -- some streams like #general you probably don't want loud @all mentions, but some streams like #ops you do.)

timabbott commented:
I think it's plausible we'd a per-stream setting for that, but I think in most organizations it's either the case that the social convenient is that `@all` mentions are used rarely for things important to everyone in the org, or not.  In any case we'd still need the default for streams where you haven't specified that explicitly, and that belongs here.

rishig commented:
hm. My prior is that you'd want at-all's "turned on" for smaller streams, and turned off for larger streams. (fwiw It's also how at least one other product ended up thinking about it.)

timabbott commented:
I think it's possible if you're in an organization that spams `@all`, you might want it for private streams only (or something).  Our approach to `@all` in larger streams is to warn people before posting (and likely in the future allow limiting use of it to stream admins), which might explain why that product handles this differently.  I favor a design where Zulip organizations generally don't have a lot of spamming of `@all` in large streams, instead of making `@all` in large streams useless (in the form of not actually notifying anyone).  

I think a per-stream toggle is super useful for that one stream that is different from your expectations around the rest, but I don't think a small-vs-large product feature around how `@all` mentions notify makes sense.

rishig commented:
Chatted with Tim -- for now the plan is to add a boolean setting 
"Suppress notifications for `@all` and `@everyone`"?
to the Notification settings page.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1033,6 +1043,18 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
+        wildcard_mention_email_user_ids = {
+            row['user_profile_id']
+            for row in subscription_rows
+            if not row['is_muted'] and row['user_profile_wildcard_mention_email_notifications']
```
With the new design, these checks should become:
```
if not row['is_muted'] and row['user_profile_wildcard_mentions_notify']
```

Though again we're probably inverting the meaning and/or wanting to only construct these lists in the event

timabbott commented:
I converted this to work as described before, and pushed back to this PR.  TODO:
* Make the `test_openapi` documentation tests pass.
* Think through and test the message editing to add/remove wildcard mentions code path(s).
* Manual testing of the web UI for this feature (totally untested)
* Open a follow-up issue for adding views and UI for the per-stream setting (data model is already done)

timabbott commented:
OK, I addressed the above issues and merged this as 1fe4f795afc5928c4a809c83823fc93425f88ec8.  

There's a couple follow-up issues I opened (#13429, #13430) that would be nice to fix afterwards and shouldn't be too much work.
